Mathlib/Topology/Algebra/Subgroup.lean

```diff
@@ -364,3 +365,219 @@ theorem isOpen_of_isOpen_subideal {U I : Ideal R} (h : U ≤ I) (hU : IsOpen (U
+  coe := fun H => H.toOpenSubgroup.toSubgroup
+
+@[to_additive]
+instance instPartialOrderOpenNormalSubgroup : PartialOrder (OpenNormalSubgroup G) := inferInstance
+
+@[to_additive]
+instance instInfOpenNormalSubgroup : Inf (OpenNormalSubgroup G) :=
+  ⟨fun U V => ⟨U.toOpenSubgroup ⊓ V.toOpenSubgroup,
+    Subgroup.normal_inf_normal U.toSubgroup V.toSubgroup⟩⟩
+
+@[to_additive]
+instance instSemilatticeInfOpenNormalSubgroup : SemilatticeInf (OpenNormalSubgroup G) :=
+  SetLike.coe_injective.semilatticeInf ((↑) : OpenNormalSubgroup G → Set G) fun _ _ => rfl
+
+@[to_additive]
+instance [ContinuousMul G] : Sup (OpenNormalSubgroup G) :=
+  ⟨fun U V => ⟨U.toOpenSubgroup ⊔ V.toOpenSubgroup,
+    Subgroup.sup_normal U.toOpenSubgroup.1 V.toOpenSubgroup.1⟩⟩
+
+@[to_additive]
+instance instSemilatticeSupOpenNormalSubgroup [ContinuousMul G] :
+    SemilatticeSup (OpenNormalSubgroup G) :=
+  toSubgroup_injective.semilatticeSup
+    (fun (H : OpenNormalSubgroup G) ↦ ↑H.toOpenSubgroup) (fun _ _ => rfl)
+
+@[to_additive]
+instance [ContinuousMul G] : Lattice (OpenNormalSubgroup G) :=
+  { instSemilatticeInfOpenNormalSubgroup,
+    instSemilatticeSupOpenNormalSubgroup with
+    toPartialOrder := instPartialOrderOpenNormalSubgroup}
+
+end OpenNormalSubgroup
+
+end
+
+/-!
+# Closed subgroups of a topological groups
+
+This section builds the SemilatticeInf `ClosedSubgroup G` of closed subgroups in a
+topological group `G`, and its additive version `ClosedAddSubgroup`.
+
+# Main definitions and results
+
+* `normalCore_isClosed` : The `normalCore` of a closed subgroup is closed.
+
+*  `finiteindex_closedSubgroup_isOpen` : A closed subgroup with finite index is open.
+
+-/
+
+section
+
+universe u v
+
+/-- The type of open subgroups of a topological group. -/
+@[ext]
+structure ClosedSubgroup (G : Type u) [Group G] [TopologicalSpace G] extends Subgroup G where
+  isClosed' : IsClosed carrier
+
+/-- The type of open subgroups of a topological group. -/
+@[ext]
+structure ClosedAddSubgroup (G : Type u) [AddGroup G] [TopologicalSpace G] extends
+    AddSubgroup G where
+  isClosed' : IsClosed carrier
+
+attribute [to_additive] ClosedSubgroup
+
+attribute [coe] ClosedSubgroup.toSubgroup ClosedAddSubgroup.toAddSubgroup
+
+namespace ClosedSubgroup
+
+variable (G : Type u) [Group G] [TopologicalSpace G]
+
+variable {G} in
+@[to_additive]
+theorem toSubgroup_injective : Function.Injective
+    (ClosedSubgroup.toSubgroup : ClosedSubgroup G → Subgroup G) :=
+  fun A B h => by
+  ext
+  rw [h]
+
+@[to_additive]
+instance : SetLike (ClosedSubgroup G) G where
+  coe U := U.1
+  coe_injective' _ _ h := toSubgroup_injective <| SetLike.ext' h
+
+@[to_additive]
+instance : SubgroupClass (ClosedSubgroup G) G where
+  mul_mem := Subsemigroup.mul_mem' _
+  one_mem U := U.one_mem'
+  inv_mem := Subgroup.inv_mem' _
+
+@[to_additive]
+instance : Coe (ClosedSubgroup G) (Subgroup G) where
+  coe := toSubgroup
+
+@[to_additive]
+instance : PartialOrder (ClosedSubgroup G) := inferInstance
```

Comment:
Surely this is unnecessary, if the proof is already `inferInstance`?

Reply:
removed